src/styles/evaluators.ts

```diff
@@ -0,0 +1,417 @@
+                        Evaluator<NumberNode>): Evaluatable<NumberNode> {
+    if (node instanceof Evaluator || node.type === 'number') {
+      return node;
+    }
+
+    switch ((node as FunctionNode).name.value) {
+      case 'calc':
+        return new CalcEvaluator(node as FunctionNode);
+      case 'env':
+        return new EnvEvaluator(node as FunctionNode);
+    }
+
+    return ZERO;
+  }
+
+  /**
+   * If the input is an Evaluator, returns the result of evaluating it.
+   * Otherwise, returns the input.
+   *
+   * This is a helper to aide in resolving a NumberNode without conditionally
+   * checking if the Evaluatable is an Evaluator everywhere.
+   */
+  static evaluate<T extends NumberNode|IdentNode>(evaluatable: Evaluatable<T>):
+      T {
+    if (evaluatable instanceof Evaluator) {
+      return evaluatable.evaluate();
+    }
+
+    return evaluatable;
+  }
+
+  /**
+   * If the input is an Evaluator, returns the value of its isConstant property.
+   * Returns true for all other input values.
+   */
+  static isConstant<T>(evaluatable: Evaluatable<T>): boolean {
+    if (evaluatable instanceof Evaluator) {
+      return evaluatable.isConstant;
+    }
+    return true;
+  }
+
+  /**
+   * If true, the Evaluator will only evaluate its AST one time. If false, the
+   * Evaluator will re-evaluate the AST each time that the public evaluate
+   * method is invoked.
+   */
+  get isConstant(): boolean {
+    return false;
+  };
+
+  protected[$lastValue]: T|null = null;
+
+  /**
+   * This method must be implemented by subclasses. Its implementation should be
+   * the actual steps to evaluate the AST, and should return the evaluated
+   * result.
+   */
+  protected abstract[$evaluate](): T;
+
+  /**
+   * Evaluate the Evaluator and return the result. If the Evaluator is constant,
+   * the corresponding AST will only be evaluated once, and the result of
+   * evaluating it the first time will be returned on all subsequent
+   * evaluations.
+   */
+  evaluate(): T {
+    if (!this.isConstant || this[$lastValue] == null) {
+      this[$lastValue] = this[$evaluate]();
+    }
+    return this[$lastValue]!;
+  }
+}
+
+
+const $identNode = Symbol('identNode');
+
+/**
+ * Evaluator for CSS-like env() functions. Currently, only one environment
+ * variable is accepted as an argument for such functions: window-scroll-y.
+ *
+ * The env() Evaluator is explicitly dynamic because it always refers to
+ * external state that changes as the user scrolls, so it should always be
+ * re-evaluated to ensure we get the most recent value.
+ *
+ * Some important notes about this feature include:
+ *
+ *  - There is no such thing as a "window-scroll-y" CSS environment variable in
+ *    any stable browser at the time that this comment is being written.
+ *  - The actual CSS env() function accepts a second argument as a fallback for
+ *    the case that the specified first argument isn't set; our syntax does not
+ *    support this second argument.
```

Comment:
A little more description here of what window-scroll-y does and how it relates to similar constructs would be nice here.

Reply:
👍